### PR TITLE
fix issue #150

### PR DESCRIPTION
**Pull request by kim-fitness**
Signed-off-by: kim-fitness <jianjin@cn.ibm.com>

This PR aims to enhance the operator to give the klusterlet an accurate condition.
It is related with open-cluster-management-io/clusteradm#65

### CODE REVIEW

File: pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go [hunk 1]
```diff
@@ -126,6 +147,39 @@ type klusterletAgent struct {
...
 
+// Check both agent deployments, if the desired replicas is equal to available replicas, return available condition
+func checkAgentsDeployment(ctx context.Context, kubeClient kubernetes.Interface, agents []klusterletAgent) metav1.Condition {
+	availableMessages := []string{}
+	for _, agent := range agents {
+		deployment, err := kubeClient.AppsV1().Deployments(agent.namespace).Get(ctx, agent.deploymentName, metav1.GetOptions{})
+		if err != nil {
+			return metav1.Condition{
+				Type:    klusterletProgressing,
+				Status:  metav1.ConditionTrue,
```
Comment (qiujian16) [inline pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go]: I think this should return Available=false

File: pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go [hunk 2]
```diff
@@ -126,6 +147,39 @@ type klusterletAgent struct {
...
+				Reason:  "GetDeploymentFailed",
+				Message: fmt.Sprintf("Failed to get deployment %q %q: %v", agent.namespace, agent.deploymentName, err),
+			}
+		}
+		if unavailablePod := helpers.NumOfUnavailablePod(deployment); unavailablePod > 0 {
+			return metav1.Condition{
```
Comment (qiujian16) [inline pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go:163]: I think you would want to also return available=true also.

**Review by qiujian16 — COMMENTED** [review]
Thanks, I think  you would want to add a integration test to check these statuses.

File: pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go (hunk 1 above)
Comment (kim-fitness) [reply pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go]: So, how we define the `Progressing` status?

File: pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go (hunk 2 above)
Comment (kim-fitness) [reply pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go:163]: OK, if the deployment is partially ready, we can also set the klusterlet condition as `Available`, let me update the code then.

**Comment by kim-fitness** [comment]
> Thanks, I think you would want to add a integration test to check these statuses.

Sorry for forgetting to append the test case, here it is.
```
1. Build the image and load it to kind cluster
2. Modify the deploy/klusterlet/config/operator/operator.yaml to set imagePullPolicy: Never
3. make deploy-spoke
4. Check the status of klusterlet as following

kubectl get klusterlets.operator.open-cluster-management.io klusterlet -o yaml
apiVersion: operator.open-cluster-management.io/v1
kind: Klusterlet
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"operator.open-cluster-management.io/v1","kind":"Klusterlet","metadata":{"annotations":{},"name":"klusterlet"},"spec":{"clusterName":"cluster1","externalServerURLs":[{"url":"https://localhost"}],"namespace":"open-cluster-management-agent","registrationImagePullSpec":"quay.io/open-cluster-management/registration:latest","workImagePullSpec":"quay.io/open-cluster-management/work:latest"}}
  creationTimestamp: "2021-09-23T05:26:29Z"
  finalizers:
  - operator.open-cluster-management.io/klusterlet-cleanup
  generation: 2
  name: klusterlet
  resourceVersion: "31342"
  uid: 32ee7b21-804c-49b3-88df-30ed1077d351
spec:
  clusterName: cluster1
  externalServerURLs:
  - url: https://localhost
  namespace: open-cluster-management-agent
  nodePlacement: {}
  registrationImagePullSpec: quay.io/open-cluster-management/registration:latest
  workImagePullSpec: quay.io/open-cluster-management/work:latest
status:
  conditions:
  - lastTransitionTime: "2021-09-23T05:26:29Z"
    message: |-
      Failed to create &SelfSubjectAccessReview{ObjectMeta:{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Spec:SelfSubjectAccessReviewSpec{ResourceAttributes:&ResourceAttributes{Namespace:,Verb:create,Group:cluster.open-cluster-management.io,Version:,Resource:managedclusters,Subresource:,Name:,},NonResourceAttributes:nil,},Status:SubjectAccessReviewStatus{Allowed:false,Reason:,EvaluationError:,Denied:false,},} with bootstrap secret "open-cluster-management-agent" "bootstrap-hub-kubeconfig": Post "https://10.96.0.1/apis/authorization.k8s.io/v1/selfsubjectaccessreviews": x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "kubernetes")
      Failed to get kubeconfig from `kubectl get secret -n "open-cluster-management-agent" "hub-kubeconfig-secret" -ojsonpath='{.data.kubeconfig}'`. This is set by the klusterlet registration deployment, but the CSR must be approved by the cluster-admin on the hub.
    reason: BootstrapSecretError,HubKubeConfigMissing
    status: "True"
    type: KlusterletRegistrationDegraded-jianjin
  - lastTransitionTime: "2021-09-23T05:26:29Z"
    message: Failed to get kubeconfig from `kubectl get secret -n "open-cluster-management-agent"
      "hub-kubeconfig-secret" -ojsonpath='{.data.kubeconfig}'`. This is set by the
      klusterlet registration deployment, but the CSR must be approved by the cluster-admin
      on the hub.
    reason: HubKubeConfigMissing
    status: "True"
    type: KlusterletWorkDegraded-jianjin
  - lastTransitionTime: "2021-09-23T05:26:29Z"
    message: 1 of requested instances are unavailable of deployment "open-cluster-management-agent"
      "klusterlet-registration-agent"
    reason: UnavailablePods
    status: "True"
    type: Progressing
  - lastTransitionTime: "2021-09-23T05:26:32Z"
    message: Klusterlet Component Applied
    reason: KlusterletApplied
    status: "True"
    type: Applied
  - lastTransitionTime: "2021-09-23T05:26:42Z"
    message: 'deployments are ready: klusterlet-registration-agent,klusterlet-work-agent'
    reason: klusterletAvailable
    status: "True"
    type: Available
  generations:
  - group: apps
    lastGeneration: 1
    name: klusterlet-registration-agent
    namespace: open-cluster-management-agent
    resource: deployments
    version: v1
  - group: apps
    lastGeneration: 1
    name: klusterlet-work-agent
    namespace: open-cluster-management-agent
    resource: deployments
    version: v1
  observedGeneration: 2
```

File: pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go (hunk 1 above)
Comment (qiujian16) [reply pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go]: progressing is a status that deployment is not fully  up, or being rolled out.

Comment (kim-fitness) [reply pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go]: Thanks for clarifying this!

File: pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go (hunk 2 above)
Comment (kim-fitness) [reply pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go:163]: > I think you would want to also return available=true also.

Hi @qiujian16 ,
Per your comment above
```
progressing is a status that deployment is not fully up, or being rolled out.
```

Here, when `unavailablePod > 0`, it means the deployment is not fully up, and should be set to condition `Progressing`, right?

**Comment by qiujian16** [comment]
@kim-fitness would you please add the test cases in https://github.com/open-cluster-management-io/registration-operator/tree/main/test/integration

**Comment by kim-fitness** [comment]
> @kim-fitness would you please add the test cases in https://github.com/open-cluster-management-io/registration-operator/tree/main/test/integration

Sure, will do then.

**Comment by kim-fitness** [comment]
I'm working on updating the unit test cases and integration test case.

**Comment by kim-fitness** [comment]
Hi @qiujian16 ,
The integration/unit test cases are ready now, and all of them have been passed successfully.
Please help to review the code then.
Thanks.

**Comment by qiujian16** [comment]
/approve
/lgtm

Thanks!

**Pull request merged**